rfd/0210-windows-ad-hsm.md

```diff
@@ -0,0 +1,198 @@
+single CRL to standard out.
+
+Teleport users setting up MS SQL access will then be able to perform a setup
+procedure very similar to what they already do, they will only have to repeat
+the `certutil -dspublish` command  once for each individual CRL.
+
+TODO: in order to publish to auth server, the CN of the CRL would need to change.
+
+#### Publishing: add CRLs to the backend `certificate_authority`
+
+In order to maintain the behavior where Teleport's `windows_desktop_service`
+publishes CRLs to LDAP on a periodic basis, we need to change the way the agent
+requests a CRL from the auth server.
+
+Today, the agent calls an auth RPC that generates a CRL on-demand. The agent has
+no control over which signer is used to generate the CRL, since it is totally
+dependent on which auth server handles the RPC.
+
+Instead of calling an RPC to get a single CRL, the agent needs to be able to get
+a set of CRLs (one for each active signer in the cluster). This means that the
+CRLs must be present somewhere in Teleport's cluster state backend and not
+generated on-demand.
+
+Since these CRLs are small (they are always empty), we will store the CRL
+directly in the `cert_authority` resource. This approach has several advantages:
+
+- Agents already set up a watch to monitor CA resources in order to implement CA rotation,
+  so the entire resource (including the CRL) should always be present in the agent's cache
+  and very cheap to look up.
+- It avoids a lot of boilerplate code required to add a new reosource and set up RBAC, caching, etc.
+
+Instead of calling an RPC to get a CRL, the agent will attempt to pull CRLs from
+the `cert_authority`. For backwards compatibliity, the agent will fall back to
+the original `Teleport` CDP if the CRL is not present in the `cert_authority`
+resource. This allows newer agents to maintain today's behavior even when running
+against older auth servers.
+
+### UX
+
+The goal of this proposal is to fix Teleport logins on HSM-enabled clusters without making
+significant changes to the user experience.
+
+#### Database access
+
+The `tctl auth crl` command will behave exactly like `tctl auth export`, printing a warning
+that asks the user to add the `--out` flag if it detects more than one active signer.
+
+We will also update the docs for this feature to ensure that it covers how to
+import multiple CRLs, and advise users to repeat the setup process when CA
+rotations are performed.
+
+#### Desktop access
+
+From a desktop access perspective, the user experience should not change at all.
+
+Users will have to update both their auth servers and their agents in order to get
+HSM support working, but we will take care to ensure that agents and auth can be
+updated independently and don't need to update in any particular order.
+
+### Proto Specification
+
+TODO: changes to `cert_authority`
+
+### Audit Events
+
+TODO: update cert.create event with signer SKID?
```

Comment:
This would be very helpful indeed. The current `cert.create` events don't mention which CA/auth server did the signing. Either SKID or auth server ID would be helpful to distinguish things.